parlai/agents/torch/torch.py

```diff
@@ -0,0 +1,257 @@
+        return shared
+
+    def vectorize(self, obs, addEndIdx=True):
+        """
+        Converts 'text' and 'label'/'eval_label' field to vectors.
+
+        :param obs: single observation from observe function
+        :param addEndIdx: default True, adds the end token to each label
+        """
+        if 'text' not in obs:
+            return obs
+        # convert 'text' field to vector using self.txt2vec and then a tensor
+        obs['text_vec'] = torch.LongTensor(self.dict.txt2vec(obs['text']))
+        if self.use_cuda:
+            obs['text'] = obs['text'].cuda()
+
+        label_type = None
+        if 'labels' in obs:
+            label_type = 'labels_vec'
+        elif 'eval_labels' in obs:
+            label_type = 'eval_labels_vec'
+
+        if label_type is not None:
+            new_labels = []
+            for label in obs[label_type[:-4]]: # leave off the _vec portion
+                vec_label = self.dict.txt2vec(label)
+                if addEndIdx:
+                    vec_label.append(self.END_IDX)
+                new_label = torch.LongTensor(vec_label)
+                if self.use_cuda:
+                    new_label = new_label.cuda()
+                new_labels.append(new_label)
+            obs[label_type] = new_labels
+        return obs
+
+    def map_valid(self, obs_batch, sort=True, is_valid=lambda obs: 'text_vec' in obs):
+        """Creates a batch of valid observations from an unchecked batch.
+        Assumes each observation has been vectorized by vectorize function.
+
+        Returns a namedtuple Batch that contains, in order, batched text
+        vectors, batched label vectors, batched labels, indices of valid
+        observations, and lengths of label vectors.
+
+        :param obs_batch: list of vectorized observations
+        :param sort: default True, orders the observations by length of vector
+        :param is_valid: default lambda, determines if an observation is valid
+        """
+        if len(obs_batch) == 0:
+            return Batch()
+
+        valid_obs = [(i, ex) for i, ex in enumerate(obs_batch) if is_valid(ex)]
+
+        if len(valid_obs) == 0:
+            return Batch()
+
+        valid_inds, exs = zip(*valid_obs)
+
+        x_text = [ex['text_vec'] for ex in exs]
+        x_lens = [ex.shape[0] for ex in x_text]
+
+        if sort:
+            ind_sorted = sorted(range(len(x_lens)), key=lambda k: -x_lens[k])
+
+            exs = [exs[k] for k in ind_sorted]
+            valid_inds = [valid_inds[k] for k in ind_sorted]
+            x_text = [x_text[k] for k in ind_sorted]
+
+        padded_xs = torch.LongTensor(len(exs),
+                                     max(x_lens)).fill_(self.NULL_IDX)
+        if self.use_cuda:
+            padded_xs = padded_xs.cuda()
+
+        for i, ex in enumerate(x_text):
+            padded_xs[i, :ex.shape[0]] = ex
+
+        xs = padded_xs
+
+        eval_labels_avail = any(['eval_labels_vec' in ex for ex in exs])
```

Comment:
having eval_labels everywhere is not good. the whole point of eval_labels was to protect models from having access to the labels.. i think we should switch to the observe(own action) thing now?

Reply:
During testing/validation don't those labels get sent through via eval_labels? What is this observe thing?